### DISCUSSION TITLE
ELI5: How this is different than https://devenv.sh/ ?

### DISCUSSION THREAD

**Discussion by barakplasma** [pkgxdev/pkgx#263] [Help]
Can I get the explain like I'm 5 version of how tea is different than https://devenv.sh/ or the Nix Package Manager?

**Comment by mxcl** [comment]
Imagine both these things but we actually work.

---

Edit: k, I'm sorry for being facetious. But you know what, when I made `brew` I got exactly the same kinds of questions. Why is this different to MacPorts, why is this different to Fink. *Why do we need this? Exactly?*

And I'm not an idiot. I get it. These questions are logical, they make sense. What I have built seems similar to everything else before, and you don't see the difference. Aren't we just reinventing the wheel?

Well obviously I don't think so. With brew, it’s proven. I didn't reinvent the wheel, I made something that was fundamentally better and with time that was clear—but in the beginning, for months, maybe even a few years that wasn't obvious and there was reasonable doubt.

The truth is it's hard to justify why something is different, why it is better. Steve Jobs said design wasn't how it looked, it was *how it worked*. I believe that.

And here that is also the case. Maybe right now that isn't obvious, but with time it will be. You can either declare me mad for this declaration or come along for the ride. We'll gladly have you along—I'm not sure Nix would be, I'm not sure devenv.sh would be. They may claim to be, but do you know how to participate? Our README is pretty clear, our codebase is pretty clear. *Our mission is pretty clear.*

How about it?

**Reply by barakplasma** [reply]
First of all, thank you for developing brew. I have been a happy user for the last 5 years, from Mac to Ubuntu. I appreciate how brew made installing environments much simpler. I have contributed to a brew formulae in the past.

As someone who heard about tea/cli yesterday, and read through the Readme and landing page for tea, some aspects of tea still aren't clear to me. I liked the section of the tea/cli readme on "vs Brew" and it answered a lot of my questions thanks to it's simple and digestible format. When evaluating which package manager ecosystem to get involved in, it helps to understand the differences between them.

Just to reflect my understanding of the differences:
||tea|devenv.sh|
|---|---|---|
|monetization|yes|no|
|decentralized|yes|no|
|reproducible envs|yes|yes|
|multi-lang packaging|yes|yes|
|simple to uninstall pkgs|yes|yes|

I'm sure I am missing nuances in the table above. I'm not trying to minimize what you're working on, just to understand in the frame of an elevator pitch / marketing sense.

I understand that the goals for tea are grand and exciting. Personally, I hope you succeed.

Feel free to delete this conversation, I didn't mean to distract you.

**Reply by olebedev** [reply]
I think the question is precisely about the technical difference, the design approach, etc. Is there a design document that describes the approach so we can clearly compare it to Nix?

**Comment by olebedev** [comment]
+1 to the initial question.

**Comment by chdiza** [comment]
> But you know what, when I made brew I got exactly the same kinds of questions. Why is this different to MacPorts, why is this different to Fink. Why do we need this? Exactly?

Those were indeed good questions, but even back then there were answers available.  Homebrew was better than MacPorts back then  because it didn't drag in oodles of gigantic dependencies for simple packages, didn't constantly rebuild things that were already built-in to macOS, made it easy to write and brew your own custom formulas, and used the "symlink farm" technique (Cellar and Prefix) which made it easy to switch versions and temporarily neutralize a package without big disk IO.  Some of those advantages were, IIRC, even explicitly advertised.

I'd love to see an updated version of this, but for tea versus devenv/nix.

**Comment by hisaac** [comment]
Just to add my 2 cents, I've been researching both `tea` and `devenv` this week as possible options for ensuring a consistent dev environment for my team. A couple thoughts:

I ran into multiple stumbling blocks and errors just when installing `devenv` and `nix` that I had to resolve on my own. It's a very complex install (at least on macOS) — in addition to moving some binaries into place on your machine, it creates a new hidden user, adds a folder to the root of your hard disk (`/nix/`), and mounts a virtual file APFS filesystem. While I'm sure there are good reasons for all that being done, it seems very heavy-handed and unnecessary for the use-case of ensuring a consistent dev/build environment. I'm certain that getting my team up and running with `devenv` would involve a lot of handholding and troubleshooting on my part, which isn't ideal.

In contrast, installing and using `tea` was super simple. One script that installs a couple things and creates some directories. Nothing more than is needed, and it was a problem-free install for me.

`nix` is definitely more mature, and the ecosystem built up around it is much more robust. And there's copious amounts of documentation, blog posts, and YouTube videos to reference. But those things just come with time, and I feel pretty confident that there will be a community and other tools built up around `tea` that will make it a great option. IMO, `tea`'s biggest issue is in clear and concise documentation, but that's only served to motivate me to want to help write some.

I also feel like `tea`'s goals of an easy and useful developer experience are more in line with what I'm looking for in this type of tool. Just like `brew` vs MacPorts, the experience of using `tea` was much more enjoyable and straightforward than using `nix`.

**Reply by olebedev** [reply]
>  it seems very heavy-handed and unnecessary for the use-case of ensuring a consistent dev/build environment.

I recommend reading through the Ph.D. thesis around Nix: https://edolstra.github.io/pubs/phd-thesis.pdf, how it ensures reproducibility, and what is essential in this process. Without understanding the fundamentals of such complex problems as package management and reproducibility, every attempt to make things like that work seems redundant/unnecessary/you name it. Ideally, I would prefer having zero movements when I decided to use something complex and having zero installation time, but we can't really ignore physics laws here, that's really frustrating.

From the light of the above, the `tea` either will replicate every building block of Nix and end up with similar complexity, at least under the hood or will be non-reproducible. This is probably not the worst, given some many package managers are not.

**Reply by mfts** [reply]
Thank you for sharing your experience with trying out `devenv` and `tea` for your team. It's a joy to hear you found `tea` simple to install and use - that's what we are aiming for. 

> IMO, tea's biggest issue is in clear and concise documentation, but that's only served to motivate me to want to help write some.

Documentation is currently spread over different places; from the various READMEs to some GitHub wiki entries. I working on compiling all documentation in one place. For now, please feel free to add to the README or a repo wiki page. Thank you.

**Comment by gettalong** [comment]
> Edit: k, I'm sorry for being facetious. But you know what, when I made brew I got exactly the same kinds of questions. Why is this different to MacPorts, why is this different to Fink. Why do we need this? Exactly?
>
> And I'm not an idiot. I get it. These questions are logical, they make sense. What I have built seems similar to everything else before, and you don't see the difference. Aren't we just reinventing the wheel?

Okay, so for me it is not the question why we need this. I see a definitive use-case. And whether there are other projects doing similar things... that's just how it is (e.g. KDE vs Gnome vs XFCE vs ...,).

It is: *What are the differences*? When would I use tea over devenv? What are the big pluses of tea vs devenv? So more of a comparison to make an informed decision. This could also be the place to strikingly show the benefits of tea!

**Comment by max-privatevoid** [comment]
I don't think pkgx can claim to "actually work" as long as it relies on libraries it didn't install.

**Reply by NotAShelf** [reply]
to quote Sun Tzu:

"lmao"
